src/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs

```diff
@@ -173,6 +160,69 @@ internal static bool IsEnabled()
 
         #region private
 
+        private const string EnableActivityPropagationEnvironmentVariableSettingName = "DOTNET_SYSTEM_NET_HTTP_ENABLEACTIVITYPROPAGATION";
+        private const string EnableActivityPropagationAppCtxSettingName = "System.Net.Http.EnableActivityPropagation";
+
+        private static readonly bool EnableActivityPropagation = GetEnableActivityPropagationValue();
+
+        private static bool GetEnableActivityPropagationValue()
+        {
+            // First check for the AppContext switch, giving it priority over the environment variable.
+            if (AppContext.TryGetSwitch(EnableActivityPropagationAppCtxSettingName, out bool enableActivityPropagation))
+            {
+                return enableActivityPropagation;
+            }
+
+            // AppContext switch wasn't used. Check the environment variable to determine which handler should be used.
+            string envVar = Environment.GetEnvironmentVariable(EnableActivityPropagationEnvironmentVariableSettingName);
+            if (envVar != null && (envVar.Equals("false", StringComparison.OrdinalIgnoreCase) || envVar.Equals("0")))
+            {
+                // Suppress Activity propagation.
+                return false;
+            }
+
+            // Defaults to enabling Activity propagation.
+            return true;
+        }
+
+        private static void InjectHeaders(Activity currentActivity, HttpRequestMessage request)
+        {
+            if (currentActivity.IdFormat == ActivityIdFormat.W3C)
+            {
+                if (!request.Headers.Contains(DiagnosticsHandlerLoggingStrings.TraceParentHeaderName))
+                {
+                    request.Headers.Add(DiagnosticsHandlerLoggingStrings.TraceParentHeaderName, currentActivity.Id);
+                    if (currentActivity.TraceStateString != null)
+                    {
+                        request.Headers.Add(DiagnosticsHandlerLoggingStrings.TraceStateHeaderName, currentActivity.TraceStateString);
+                    }
+                }
+            }
+            else
+            {
+                if (!request.Headers.Contains(DiagnosticsHandlerLoggingStrings.RequestIdHeaderName))
+                {
+                    request.Headers.Add(DiagnosticsHandlerLoggingStrings.RequestIdHeaderName, currentActivity.Id);
+                }
+            }
+
+            // we expect baggage to be empty or contain a few items
+            using (IEnumerator<KeyValuePair<string, string>> e = currentActivity.Baggage.GetEnumerator())
```

Comment:
The way Baggage is implemented, this is always going to allocate, even if it's empty. I suggest Baggage be fixed to only allocate an enumerable/enumerator if it's not empty.

Reply:
https://github.com/dotnet/corefx/pull/38389